utilities/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+extern crate geographiclib_rs;
+
+use geographiclib_rs::Geodesic;
+use std::collections::HashMap;
+use std::error::Error;
+use std::fs::{self, File};
+use std::io::{self, BufRead, prelude::*};
+use std::path;
+use std::time;
+use std::u64;
+use zip;
+
+// todo: add a rust unit test that mirrors C++ MGRS::Check
+
+const ZIP_PATH_RELATIVE: &str = "test_fixtures/geographiclib-instrumented/geographiclib-instrumented-crude-out";
+const DAT_PATH_RELATIVE: &str = "test_fixtures/test-data-unzipped";
+
+#[allow(non_upper_case_globals)]
+pub const nC_: usize = 7; // todo: define in geodesic-rs instead
+
+pub struct DataPathPair {
+    pub path_zip: path::PathBuf,
+    pub path_dat: path::PathBuf,
+}
+
+// Given an operation name, return the path to the zipped and unzipped dat file for that operation.
+// Note that the unzipped path may point to a location that doesn't exist.
+pub fn get_data_paths(op_name: &str) -> std::io::Result<DataPathPair> {
+    let mut filename_zip = op_name.to_owned();
+    let mut filename_dat = op_name.to_owned();
+    filename_zip.push_str(".dat.zip");
+    filename_dat.push_str(".dat");
+
+    let dir_base = std::env::current_dir()?;
+    let path_base = dir_base.as_path();
+    let path_zip = path::Path::new(path_base).join(ZIP_PATH_RELATIVE);
+    let path_dat = path::Path::new(path_base).join(DAT_PATH_RELATIVE);
+    let result = DataPathPair {
+        path_zip: path_zip.join(filename_zip),
+        path_dat: path_dat.join(filename_dat),
+    };
+    Ok(result)
+}
+
+pub fn refresh_unzipped(paths: &DataPathPair) -> std::io::Result<()> {
```

Comment:
What is this for?

Reply:
My instrumented C++ repo currently stores zipped copies of the data files, to cut down on volume. Over time, I've gotten a little less sure whether this zipping is actually very helpful, because I'm pretty sure that git compresses data before sending and storing. It's possible that I just got unlucky with sourceforge hanging up on me the first few times I tried storing the raw data files.

In any event, those zipped data files are pulled from my c++ repo as part of a "submodule" under the test_fixtures folder. `refresh_unzipped` takes a file base name, looks for the zipped and unzipped copies of the file, compares their timestamps, and does a fresh unzip if the unzipped version either doesn't exist or is older than the zipped version. I can certainly beef up the function-level comments in the utilities lib file to discuss this sort of thing.